### PR TITLE
Filter large responses

### PR DESCRIPTION
**Pull request by kivancmuslu**

### CODE REVIEW

**Comment by kivancmuslu** [comment]
Actual filtering and telemetry will be done in the ADO collectors based on the return value of Try... method (new method).

File: Core.Collectors/Web/HttpUtility.cs [hunk 1]
```diff
@@ -19,6 +22,22 @@ public static async Task<JObject> ParseAsJObjectAsync(HttpResponseMessage respon
             return JObject.Load(jsonTextReader);
         }
 
+        public static async Task<Tuple<bool, JObject>> TryParseAsJObjectAsync(HttpResponseMessage response)
+        {
+            using Stream responseStream = await response.Content.ReadAsStreamAsync().ConfigureAwait(false);
+            // Loading a stream / text into JObject expands ~10x in memory. In other words, a 100MB response, will approx. take 1GB ram.
+            // Our machines (where this is a problem --- for CG collection) have 16 GB ram. I am leaving 4 GB for OS (might not be needed but just in case) and the rest of the code.
+            // That leaves 12 GB to be shared between 6, or 2 GB each, so ~200 MB raw response.
+            if (responseStream.Length > 200 * MB)
```
Comment (maxi) [inline Core.Collectors/Web/HttpUtility.cs:31]: Did you confirm that this actually works? Is the response always fully buffered and the `Length` property valid and correct? If it's not fully buffered, and it's a streaming response without a Content-Length header, what is the value here?
0, -1, int.MaxValue, some arbitrary buffer size?

Comment (maxi) [inline Core.Collectors/Web/HttpUtility.cs:31]: What about making the max size a default parameter to the function?

File: Core.Collectors/Web/HttpUtility.cs [hunk 2]
```diff
@@ -19,6 +22,22 @@ public static async Task<JObject> ParseAsJObjectAsync(HttpResponseMessage respon
             return JObject.Load(jsonTextReader);
         }
 
+        public static async Task<Tuple<bool, JObject>> TryParseAsJObjectAsync(HttpResponseMessage response)
```
[the remark this answers is not in the document]
Comment (kivancmuslu) [reply Core.Collectors/Web/HttpUtility.cs:25]: out parameter does not work due to async but I can try out nullable type, thanks for suggestion.

[the remark this answers is not in the document]
Comment (maxi) [reply Core.Collectors/Web/HttpUtility.cs:25]: Oh, of course. Good point. 🤦‍♂️

File: Core.Collectors/Web/HttpUtility.cs (hunk 1 above)
Comment (kivancmuslu) [reply Core.Collectors/Web/HttpUtility.cs:31]: What do you have in mind? Do you want to also make this customizable at the collection node level?

Comment (maxi) [reply Core.Collectors/Web/HttpUtility.cs:31]: Not much actually. It was just a thought, why have this limit hardcoded in the function? Making this a parameter makes it easier to change if we need to -- let's say, we need a 250 MiB limit or 200 is already too much. No need to mess with the submodule then.

Comment (kivancmuslu) [reply Core.Collectors/Web/HttpUtility.cs:31]: Fair point, did not think about hardcoding in submodule vs. making it easily change-able on the caller end. WIll change.

Comment (maxi) [reply Core.Collectors/Web/HttpUtility.cs]: The response seems to be always buffered. I think we can go with this.

**Review by maxi — APPROVED** [review]

**Review by AllainPL — APPROVED** [review]

**Pull request merged**